---
title: "The RNA Centric Analysis System Report"
author: "Bora Uyar, Dilmurat Yusuf, Ricardo Wurmus, Altuna Akalin"
date: "`r Sys.Date()`"
bibliography: bibliography.bib
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    theme: simplex
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r load_libraries, results='hide', message=FALSE, warning=FALSE, echo=FALSE}
library(RCAS)
library(data.table)
library(plotly)
library(DT)
library(motifRG)
```

# Introduction

RNA Centric Annotation System (RCAS) automatically generates
dynamic annotations for custom input files
that contain transcriptomic target regions.
High-throughput sequencing (HTS) has brought a profound impact
on biological research at various levels.
At the level of transcriptome,
various HTS methods have been applied to study different aspects of
the complete set of transcripts that are produced by the genome.
On such method is
cross-linking immunoprecipitation-high-throughput
sequencing (CLIP-Seq) [@Konig2011]
that is used to
study RNA binding sites for a protein of interest on a genome-wide scale,
thereby increasing the understanding of
post-transcriptional regulatory networks.
To mitigate the challenge posed by
big data generated by HTS methods,
automatic solution, such RCAS, is required to
to facilitate biological discovery
from taget regions located by methods such as Clip-Seq.
RCAS automatically provides
dynamic annotations for custom input files
that contain transcriptomic target regions.
Provided with target regions in
[BED](http://www.ensembl.org/info/website/upload/bed.html) format(
and annotation references in
[GFF/GTF](http://www.ensembl.org/info/website/upload/gff.html) format,
RCAS automatically summarizes annotated features that overlap with targets,
and perform enrichment analysis of
motifs, Gene Ontology (GO) terms and gene set.
The summary of annotated features
consists of intuitive reports in terms of
distributions of gene types, genomic features, gene names
and coverage profiles.
Motif discovery is also one key component of RCAS automatic system.
RCAS can provide users list of motifs
that are enriched comparing with background sequences.
In addition,
RCAS generates lists of enriched gene sets and functions
to shed light on the functional aspects of target regions.
The final report of RCAS consists of high-quality dynamic figures and tables,
which are readily applicable for publications or other academic usage.


#  Data input

RCAS requires inputs of a BED file and a GFF/GTF file.
The BED file should contain coordinates/intervals
of target transcriptomic regions
which are located via transcriptomics methods such as Clip-Seq.
The GFF/GTF file should provide reference annotation.
The usual source of GFF/GTF file are
[ENSEMBLE](http://www.ensembl.org/info/data/ftp/index.html) and
[GENCODE](http://www.gencodegenes.org/releases/).
Meanwhile user can also provide custom GFF/GTF file.

For this vignette.
to demonstrate RCAS functionality,
we  provide sample BED and GFF file via RCAS library,
which can be imported using a common R function, data().
To import custom BED and GFF/GTF files,
user should execute two RCAS functions called importBed() and importGtf().

## Importing sample data
```{r sample_data, warning=FALSE, message=FALSE}
library(RCAS)
data(queryRegions) #sample queryRegions in BED format()
data(gff)          #sample GFF file
```

## Importing custom data
To use importBed() and importGtf(),
the user should provide file paths to the respective
BED file and GFF/GTF file.
To reduce memory usage and time consumption,
we advise user to set `sampleN=10000` to avoid
huge input of intervals.

Before running following codes in your R interpreter,
please remove `#` comment sign.
```{r RCAS_import_data, warning=FALSE, message=FALSE}
#library(RCAS)
#queryRegions <- importBed(filePath = BED_file, sampleN = 10000)
#gff <- importGtf(filePath = GFF_file)
```

# Summarizing Overlaps of Query Regions with Genomic Annotation Features

## Querying the annotation file 
```{r queryGFF, warning=FALSE, message=FALSE}
overlaps <- queryGff(queryRegions = queryRegions, gff = gff)
#data.table is used to do quick summary operations
overlaps.dt <- data.table(as.data.frame(overlaps)) 
```


### Finding targeted gene types
To find out the distribution of the query regions across gene types:
```{r query_gene_types, warning=FALSE, message=FALSE}
biotype_col <- grep('biotype', colnames(overlaps.dt), value = T)
df <- overlaps.dt[,length(unique(overlappingQuery)), by = biotype_col]
colnames(df) <- c("feature", "count")
df$percent <- round(df$count / length(queryRegions) * 100, 1)
df <- df[order(count, decreasing = TRUE)]
p <- plot_ly(data = df, 
             type = "bar",
             x = feature,
             y = percent,
             text = paste("count:", count), color=feature)
layout(p = p, 
       margin = list(l=100, r=100, b=150), 
       xaxis = list(showticklabels = TRUE,  tickangle = 90), 
       yaxis = list(title = paste("percentage of query regions,", 
                                  "n =", length(queryRegions))))
```

### Distribution of query regions across chromosomes
To find out the distribution of the query regions across chromosomes:
```{r chromosomes_gene_features, warning=FALSE, message=FALSE}
df <- overlaps.dt[,length(unique(overlappingQuery)), by = c('seqnames', 'type')]
colnames(df) <- c('seqnames', 'type', 'count')
df <- df[order(seqnames)]
p <- plot_ly(data = df, 
             type = "bar", 
             x = seqnames,
             y = count, 
             text = paste("count:", count), 
             color = type)
layout(p = p, 
       margin = list(b=150), 
       xaxis = list(showticklabels = TRUE,  tickangle = 90), 
       yaxis = list(title = paste("Number of query regions,", 
                                  "n =", length(queryRegions)
                                  )
                    )
       )
```

## Extending the annotation feature space 

GTF files contain some annotation features (e.g. exons, transcripts) that are 
usually explicitly defined, however, some gene features such as introns,
exon-intron boundaries, promoter regions are only implicitly defined. Such 
implicity features can be extracted from a GTF file using makeTxDb family of 
functions from the GenomicFeatures library.

First we create a list of GRanges objects, where each list element contains all
the available coordinates of gene features such as transcripts, exons, introns,
5'/3' UTRs, exon-intron boundaries, and promoter regions. 

```{r getTxdbFeatures, warning=FALSE, message=FALSE}
txdbFeatures <- getTxdbFeaturesFromGff(gff)
```

### Plotting overlap counts between query regions and gene features
To have a global overview of the distribution of query regions across gene
features, we can use the summarizeQueryRegions function.

```{R summarizeQueryRegions, warning=FALSE, message=FALSE}
summary <- summarizeQueryRegions(queryRegions = queryRegions, 
                                 txdbFeatures = txdbFeatures)

df <- data.frame(summary)
df$percent <- round((df$count / length(queryRegions)), 3) * 100
p <- plot_ly( data = df, 
              x = rownames(df), 
              y = percent, 
              type = 'bar',
              text = paste("count:", count), 
              color = rownames(df)
              )
layout(p = p, 
       xaxis = list(title = 'features'),
       yaxis = list(title = paste("percentage of query regions,", 
                                  "n =", length(queryRegions)
                                  )
                    ), 
       margin = list(b = 150, r = 50)
       )
```

### Obtaining a table of overlap counts between query regions and genes 
To find out which genes overlap with how many queries and categorise overlaps by
gene features; we use getTargetedGenesTable function, which returns a data.frame
object. Then we use datatable function from 'DT' library to print an interactive
table of genes that overlap query Regions.
```{r getTargetedGenesTable, warning=FALSE, message=FALSE}
dt = getTargetedGenesTable(queryRegions = queryRegions, 
                           txdbFeatures = txdbFeatures)
datatable(data = dt[order(transcripts, decreasing = T)][1:500], 
          extensions = 'FixedColumns', 
          options = list(
            dom = 't',
            scrollX = TRUE, 
            scrollCollapse = TRUE
            )
          )
```

### Profiling the coverage of query regions across gene features 

#### Coverage profile of query regions for 3' UTRs
Coverage profiles can be obtained for a single type of gene feature or a list of
gene features. Here we demonstrate how to get coverage profile of query regions
across 3'UTRs. It might be a good idea to use sampleN parameter to randomly
downsample the target regions to speed up the calculations.

```{r coverageprofile, warning=FALSE, message=FALSE}
cov <- calculateCoverageProfile(queryRegions = queryRegions, 
                               targetRegions = txdbFeatures$threeUTRs, 
                               sampleN = 10000)

p <- plot_ly(cov, x = bins, y = coverage)
p %>%   add_trace(y = fitted(loess(coverage ~ as.numeric(bins)))) %>%
layout(title = paste("Coverage along 3'UTRs (5' -> 3' direction)", sep = " "), 
       showlegend = FALSE, 
       margin = list(l = 50, r = 50, b = 50, t = 50))

```

#### Coverage profile of query regions for all gene features
Coverage profiles can be obtained for a single type of gene feature or a list of
gene features. Here we demonstrate how to get coverage profile of query regions 
across all available gene features. It might be a good idea to use sampleN
parameter to randomly downsample the target regions to speed up the 
calculations.

```{r coverageprofilelist, warning=FALSE, message=FALSE}
covList <- calculateCoverageProfileList(queryRegions = queryRegions, 
                                       targetRegionsList = txdbFeatures, 
                                       sampleN = 10000)

df <- do.call('cbind', covList)
df <- df[,!grepl(colnames(df), pattern = '*.bins')]
df$bins <- c(1:100)
colnames(df) <- gsub(pattern = ".coverage", replacement = "", x = colnames(df))
mdf <- reshape2::melt(df, id.vars = c('bins'))
colnames(mdf) <- c('bins', 'feature', 'coverage')
p = plot_ly(data = mdf, x = bins, y = coverage, color = feature)
layout(p)
```

# Motif Analysis using motifRG

## Calculating enriched motifs
With the RCAS package, a motif analysis is also possible.
RCAS uses motifRG library to find enriched motifs among the query regions.
```{r motif_analysis, warning=FALSE, message=FALSE}
motifResults <- runMotifRG(queryRegions = queryRegions, 
                           genomeVersion = 'hg19', 
                           motifN = 4)

par(mfrow = c(2,2), mar = c(2,2,2,2))
for (i in 1:length(motifResults$motifs)) {
  motifPattern <- motifResults$motifs[[i]]@pattern
  motifRG::plotMotif(match = motifResults$motifs[[i]]@match$pattern, 
                     main = paste0('Motif-',i,': ',motifPattern),
                     entropy = T)
}
```

## motif analysis: getting motif summary statistics
A summary table from the motif analysis results can be obtained
```{r motif_analysis_table, warning=FALSE, message=FALSE}
summary <- getMotifSummaryTable(motifResults)
datatable(summary, extensions = 'FixedColumns', 
          options = list( 
            dom = 't', 
            scrollX = TRUE, 
            scrollCollapse = TRUE
            )
          )
```


# GO term analysis

## Biological processes enriched among targeted genes
RCAS can perform GO term enrichment analysis to find out
enriched functions in genes that overlap the query regions
```{r GO analysis, warning=FALSE, message=FALSE}

#get all genes from the GTF data
backgroundGenes <- unique(gff$gene_id)
#get genes that overlap query regions
targetedGenes <- unique(overlaps$gene_id)

#run TopGO
goResults <- runTopGO(ontology = 'BP', 
                      species = 'human', 
                      backgroundGenes = backgroundGenes, 
                      targetedGenes = targetedGenes)

datatable(data = goResults[goResults$bh < 0.1,], 
          extensions = 'FixedColumns', 
          options = list(dom = 't',
                         scrollX = TRUE,
                         scrollCollapse = TRUE
                         )
          )

```


# Gene Set Enrichment Analysis

## MSIGDB gene sets enriched among targeted genes
RCAS can use gene sets from Molecular Signatures Database
and calculate gene set enrichment analysis
to find out which gene sets are enriched
among the genes targeted by the query regions
```{r msigdb_analysis, warning=FALSE, message=FALSE}
#msigDB <- parseMsigdb(msigdbFile)
data(msigDB)
msigdbResults <- runMSIGDB(msigDB = msigDB,
    backgroundGenes = backgroundGenes, targetedGenes = targetedGenes)

datatable(msigdbResults[msigdbResults$BH < 0.1,],
    extensions = 'FixedColumns',
  options = list(
    dom = 't',
    scrollX = TRUE,
    scrollCollapse = TRUE
  ))

```


# Acknowledgements

RCAS is developed by
[Dr. Altuna Akalin](http://bioinformatics.mdc-berlin.de/team.html#altuna-akalin-phd)
(head of the Scientific Bioinformatics Platform),
[Dr. Bora Uyar](http://bioinformatics.mdc-berlin.de/team.html#bora-uyar-phd)
(Bioinformatics Scientist),
[Dr. Dilmurat Yusuf](http://bioinformatics.mdc-berlin.de/team.html#dilmurat-yusuf-phd)
(Bioinformatics Scientist) and 
[Ricardo Wurmus](http://bioinformatics.mdc-berlin.de/team.html#ricardo-wurmus)
(System Administrator) at the Berlin Institute of Medical Systems Biology
([BIMSB](https://www.mdc-berlin.de/13800178/en/bimsb))
at the Max-Delbrueck-Center for Molecular Medicine
([MDC](https://www.mdc-berlin.de)) in Berlin.

RCAS is developed as a bioinformatics service as part of
the [RNA Bioinformatics Center](http://www.denbi.de/index.php/rbc),
which is one of the eight centers of
the German Network for Bioinformatics Infrastructure
([de.NBI](http://www.denbi.de/)).  

# References
