---
title: "The RNA Centric Analysis System Report"
author: "Bora Uyar"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    theme: simplex
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r load_libraries, results='hide', message=FALSE, warning=FALSE, echo=FALSE}
library(RCAS)
library(data.table)
library(plotly)
library(DT)
library(motifRG)
```

# Introduction
RNA Centric Annotation System is an automated system that provides
dynamic annotations for custom input files that contain transcriptomic target
regions. Such transcriptomic target regions could be, for instance, peak regions
detected by CLIP-Seq analysis that detect protein-RNA interactions, MeRIP-Seq
analysis that detect RNA modifications (alias the epitranscriptome), or any
collection of target regions at the level of the transcriptome. RCAS contains
wrapper functions to do de novo motif discovery using functions from the motifRG
package. RCAS overlays the input target regions with the annotated protein-
coding genes and calculates the Gene Ontology (GO) terms that may be enriched or
depleted in the input target regions compared to the background list of protein-
coding genes. A Classical Fisher's Exact Test is applied for each GO term and
the p-values obtained for each GO term is corrected for multiple testing using
both the False Discovery Rate and the Family-Wise Error Rate. Similarly to the
GO term enrichment analysis, RCAS also detects sets of genes as annotated in
the Molecular Signatures Database that are enriched or depleted in the queried
target regions. Results are corrected for multiple-testing according to both the
False Discovery Rate and the Family-Wise Error Rate.

```{r get_inputs, warning=FALSE, message=FALSE}
peaksFile <- '/Users/buyar/Desktop/data/bed/HITSCLIP_LIN28AWilbert2012a_hg19.bed'
gtfFile <- '/Users/buyar/Desktop/data/gff/hg19/Homo_sapiens.GRCh37.75.gtf'
msigdbFile <- '/Users/buyar/RCAS/src/base/c2.cp.v5.0.entrez.hg19.gmt'
```

# Getting Input

## Importing GTF files

Import the contents of the input GTF file as a GRanges object. 
```{r importGtf, warning=FALSE, message=FALSE}
gff = importGtf(filePath = gtfFile)
```


## Importing query regions from BED files 
Import the contents of the input BED file as a GRanges object. Use sampleN=<integer> to randomly down-sample the number of intervals so that the analysis doesn't take too much memory and time in case the input contains too many intervals. A reasonable value for this parameter (sampleN) is 10000.  
```{r importBed, warning=FALSE, message=FALSE}
queryRegions = importBed(filePath = peaksFile, sampleN = 10000)
```

# Summarizing Overlaps

## Querying the GFF 
Given the query regions from BED file and the genomic annotation from the GTF file, we can do some overlap operations to find out the annotation properties of the input query regions. 
```{r queryGFF, warning=FALSE, message=FALSE}
overlaps = queryGff(queryRegions = queryRegions, gff = gff)
overlaps.dt = data.table(as.data.frame(overlaps)) #data.table is used to do quick summary operations 
```


## Querying the GFF: targeted gene types 
To find out the distribution of the query regions across gene types:
```{r query_gene_types, warning=FALSE, message=FALSE}
biotype_col = grep('biotype', colnames(overlaps.dt), value=T)
df = overlaps.dt[,length(unique(overlappingQuery)), by=biotype_col]
colnames(df) = c("feature", "count")
df$percent = round(df$count / length(queryRegions) * 100, 1)
df = df[order(count, decreasing = TRUE)]
p = plot_ly(df, type="bar", x = feature, y = percent, text=paste("count:", count), color=feature)
layout(p, margin = list(l=100, r=100, b=150), xaxis = list(showticklabels = TRUE,  tickangle = 90), yaxis = list(title = paste("percentage of query regions,", "n =",length(queryRegions))))
```

## Querying the GFF: targeted chromosomes 
To find out the distribution of the query regions across chromosomes:
```{r chromosomes_gene_features, warning=FALSE, message=FALSE}
df = overlaps.dt[,length(unique(overlappingQuery)), by=c('seqnames', 'type')]
colnames(df) = c('seqnames', 'type', 'count')
df = df[order(seqnames)]
p = plot_ly(df, type="bar", x = seqnames, y = count, text=paste("count:", count), color=type)
layout(p, margin = list(b=150), xaxis = list(showticklabels = TRUE,  tickangle = 90), yaxis = list(title = paste("Number of query regions,", "n =",length(queryRegions))))
```


# Querying TxDB
Here we use the makeTxDb functions of GenomicFeatures library to find out about the distribution of query regions across gene features. 

## Querying TxDB: creating txdbFeatures 
First we create a list of GRanges objects, where each list element contains all the available coordinates of gene features such as transcripts, exons, introns, 5'/3' UTRs, exon-intron boundaries, and promoter regions. 
```{r getTxdbFeatures, warning=FALSE, message=FALSE}
txdbFeatures = getTxdbFeaturesFromGff(gff)
```

## Querying TxDB: targeted genes table 
To find out which genes overlap with how many queries and categorise overlaps by gene features; we use getTargetedGenesTable function, which returns a data.frame object. Then we use datatable function from 'DT' library to print an interactive table of genes that overlap query Regions.   
```{r getTargetedGenesTable, warning=FALSE, message=FALSE}
dt = getTargetedGenesTable(queryRegions = queryRegions, txdbFeatures = txdbFeatures)
datatable(dt[order(transcripts, decreasing = T)][1:500], extensions = 'FixedColumns',
  options = list(
    dom = 't',
    scrollX = TRUE,
    scrollCollapse = TRUE
  ))
```

## Querying TxDB: Coverage Profiles

### Querying TxDB: Coverage Profile for Exons
To calculate the coverage profiles of query Regions across gene features. It might be a good idea to use sampleN parameter to randomly downsample the target regions to speed up the calculations. 
```{r coverageprofile, warning=FALSE, message=FALSE}
cov = calculateCoverageProfile(queryRegions = queryRegions, targetRegions = txdbFeatures$threeUTRs, sampleN = 10000)

p = plot_ly(cov, x = bins, y = coverage)
p %>%  
add_trace(y = fitted(loess(coverage ~ as.numeric(bins)))) %>%
layout(title = paste("Coverage along 3'UTRs (5' -> 3' direction)", sep=" "), showlegend = FALSE, margin = list(l= 50, r=50, b=50, t=50))

```

### Querying TxDB: Coverage profile for all gene features 
To calculate the coverage profiles of query regions across all possible types of gene features. It might be a good idea to use sampleN parameter to randomly downsample the target regions to speed up the calculations. 
```{r coverageprofilelist, warning=FALSE, message=FALSE}
covList = calculateCoverageProfileList(queryRegions = queryRegions, targetRegionsList = txdbFeatures, sampleN = 10000)

df = do.call('cbind', covList)
df = df[,!grepl(colnames(df), pattern = '*.bins')]
df$bins = c(1:100)
colnames(df) = gsub(pattern = ".coverage", replacement = "", x = colnames(df))
mdf = reshape2::melt(df, id.vars = c('bins'))
colnames(mdf) = c('bins', 'feature', 'coverage')
p = plot_ly(data = mdf, x=bins, y=coverage, color = feature)
layout(p)
```

# Motif Analysis using motifRG

## Calculating enriched motifs
With the RCAS package, a motif analysis is also possible. RCAS uses motifRG library to find enriched motifs among the query regions. 
```{r motif_analysis, warning=FALSE, message=FALSE}
motifResults <- runMotifRG(queryRegions = queryRegions, genomeVersion = 'hg19', motifN = 3)

par(mfrow=c(2,2), mar=c(2,2,2,2))
for (i in 1:length(motifResults$motifs)){
  motifPattern = motifResults$motifs[[i]]@pattern
  motifRG::plotMotif(motifResults$motifs[[i]]@match$pattern, main=paste0('Motif-',i,': ',motifPattern), entropy = T)
}

```

## motif analysis: getting motif summary statistics
A summary table from the motif analysis results can be obtained
```{r motif_analysis_table, warning=FALSE, message=FALSE}
summary <- getMotifSummaryTable(motifResults)
datatable(summary, extensions = 'FixedColumns',
  options = list(
    dom = 't',
    scrollX = TRUE,
    scrollCollapse = TRUE
  ))
```


# GO term analysis 

## Biological processes enriched among targeted genes 
RCAS can perform GO term enrichment analysis to find out enriched functions in genes that overlap the query regions 
```{r GO analysis, warning=FALSE, message=FALSE}

#get all genes from the gff data
backgroundGenes = unique(gff$gene_id)
#get genes that overlap query regions
targetedGenes = unique(overlaps$gene_id)

#run TopGO 
goResults = runTopGO(ontology = 'BP', species = 'human', backgroundGenes = backgroundGenes, targetedGenes = targetedGenes)

datatable(goResults[goResults$bh < 0.1,], extensions = 'FixedColumns',
  options = list(
    dom = 't',
    scrollX = TRUE,
    scrollCollapse = TRUE
  ))

```


# Gene Set Enrichment Analysis 

## MSIGDB gene sets enriched among targeted genes 
RCAS can use gene sets from Molecular Signatures Database and calculate gene set enrichment analysis to find out which gene sets are enriched among the genes targeted by the query regions 
```{r msigdb_analysis, warning=FALSE, message=FALSE}
msigDB <- parseMsigdb(msigdbFile) 
msigdbResults <- runMSIGDB(msigDB = msigDB, backgroundGenes = backgroundGenes, targetedGenes = targetedGenes)

datatable(msigdbResults[msigdbResults$BH < 0.1,], extensions = 'FixedColumns',
  options = list(
    dom = 't',
    scrollX = TRUE,
    scrollCollapse = TRUE
  ))

```


# Acknowledgements

RCAS is developed by [Dr. Altuna Akalin](http://bioinformatics.mdc-berlin.de/team.html#altuna-akalin-phd) (head of the Scientific Bioinformatics Platform), [Dr. Dilmurat Yusuf](http://bioinformatics.mdc-berlin.de/team.html#dilmurat-yusuf-phd) (Bioinformatics Scientist), [Dr. Bora Uyar](http://bioinformatics.mdc-berlin.de/team.html#bora-uyar-phd)  (Bioinformatics Scientist), and [Ricardo Wurmus](http://bioinformatics.mdc-berlin.de/team.html#ricardo-wurmus) (System Administrator) at the Berlin Institute of Medical Systems Biology ([BIMSB](https://www.mdc-berlin.de/13800178/en/bimsb)) at the Max-Delbrueck-Center for Molecular Medicine ([MDC](https://www.mdc-berlin.de)) in Berlin. 

RCAS is developed as a bioinformatics service as part of the [RNA Bioinformatics Center](http://www.denbi.de/index.php/rbc), which is one of the eight centers of the German Network for Bioinformatics Infrastructure ([de.NBI](http://www.denbi.de/)).  

