---
title: "The RNA Centric Analysis System Report"
author: "Bora Uyar, Dilmurat Yusuf, Ricardo Wurmus, Altuna Akalin"
date: "`r Sys.Date()`"
params: 
  query: 'testdata'
  gff: 'testdata'
  msigdb: 'testdata'
  annotationSummary: TRUE
  goAnalysis: TRUE
  msigdbAnalysis: TRUE
  motifAnalysis: TRUE
  genomeVersion: 'hg19'
  species: 'human'
  printProcessedTables: FALSE
  sampleN: 0
  workdir: '.'
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
knitr::opts_knit$set(root.dir = params$workdir)
```

```{r chunk_eval_options}
#if params$motifAnalysis is TRUE and any motif is found, this will set to TRUE
provideMotifSummary <- FALSE 
goSummaryBP <- FALSE 
goSummaryMF <- FALSE
goSummaryCC <- FALSE 
```

```{r load_libraries, results='hide'}
library(RCAS)
```

# Introduction

RNA Centric Annotation System (RCAS) automatically generates
dynamic annotations for custom input files
that contain transcriptomic target regions.
High-throughput sequencing (HTS) has brought a profound impact
on biological research at various levels.
At the level of transcriptome,
various HTS methods have been applied to study different aspects of
the complete set of transcripts that are produced by the genome.
One such method is
cross-linking immunoprecipitation-high-throughput
sequencing (CLIP-Seq)
that is used to
study RNA binding sites for a protein of interest on a genome-wide scale,
thereby increasing the understanding of
post-transcriptional regulatory networks.
To mitigate the challenge posed by
big data generated by HTS methods,
automatic solution, such RCAS, is required to
to facilitate biological discovery
from taget regions located by methods such as Clip-Seq.
RCAS automatically provides
dynamic annotations for custom input files
that contain transcriptomic target regions.
Provided with target regions in
[BED](http://www.ensembl.org/info/website/upload/bed.html) format(
and annotation references in
[GFF/GTF](http://www.ensembl.org/info/website/upload/gff.html) format,
RCAS automatically summarizes annotated features that overlap with targets,
and perform enrichment analysis of
motifs, Gene Ontology (GO) terms and gene set.
The summary of annotated features
consists of intuitive reports in terms of
distributions of gene types, genomic features, gene names
and coverage profiles.
Motif discovery is also one key component of RCAS automatic system.
RCAS can provide users list of motifs
that are enriched comparing with background sequences.
In addition,
RCAS generates lists of enriched gene sets and functions
to shed light on the functional aspects of target regions.
The final report of RCAS consists of high-quality dynamic figures and tables,
which are readily applicable for publications or other academic usage.

```{r initiate_figure_table_counts}
figureCount <- 1
tableCount <- 1
```

```{r getInput, results='hide'}
if(params$query == 'testdata') {
  data(queryRegions) 
} else {
  cat('importing BED from filepath',params$query,'\n')
  queryRegions <- importBed(filePath = params$query, sampleN = params$sampleN)
}

if(params$gff == 'testdata') {
  data(gff)         
} else {
  cat('importing GTF from filepath',params$gff,'\n')
  gff <- importGtf(filePath = params$gff)
}

overlaps <- queryGff(queryRegions = queryRegions, gff = gff)
#data.table is used to do quick summary operations
overlaps.dt <- data.table(GenomicRanges::as.data.frame(overlaps)) 

#get all genes from the GTF data 
backgroundGenes <- unique(gff$gene_id)
#get genes that overlap query regions
targetedGenes <- unique(overlaps$gene_id)

```

```{r getTxdbFeatures, eval=params$annotationSummary}
txdbFeatures <- getTxdbFeaturesFromGff(gff)
```

```{r summaryOverlapsHeader, results='asis', eval=params$annotationSummary}
cat('# Annotation Summary for Query Regions \n')
```

```{r summarizeQueryRegionsFigureCaption, results='asis', eval=params$annotationSummary}
cat('## Distribution of query regions across gene features\n')

cat("**Figure",figureCount,":** The number of query regions that overlap different kinds of gene features are counted. The 'y' axis denotes the types of gene features included in the analysis and the 'x' axis denotes the percentage of query regions (out of total number of query regions denoted with 'n') that overlap at least one genomic interval that host the corresponding feature. Notice that the sum of the percentage values for different features don't add up to 100%, because some query regions may overlap multiple kinds of features \n")
```


```{r summarizeQueryRegions, eval=params$annotationSummary}
summary <- summarizeQueryRegions(queryRegions = queryRegions, 
                                 txdbFeatures = txdbFeatures)

df <- data.frame(summary)
df$percent <- round((df$count / length(queryRegions)), 3) * 100
p <- plot_ly( data = df, 
              x = rownames(df), 
              y = percent, 
              type = 'bar',
              text = paste("count:", count), 
              color = rownames(df)
              )
layout(p = p, 
       xaxis = list(title = 'features'),
       yaxis = list(title = paste("percentage of query regions,", 
                                  "n =", length(queryRegions)
                                  )
                    ), 
       margin = list(b = 150, r = 50)
       )
figureCount <- figureCount + 1
```

```{r get_table_of_genes_tabcap, echo=FALSE, results='asis', eval=params$annotationSummary}
cat("## Interactive table of genes that overlap query regions\n")
cat("**Table",tableCount,":** Interactive table of top 100 genes that overlap query regions, grouped by gene features such as introns, exons, UTRs, etc.\n")
```

```{r getTargetedGenesTable, eval=params$annotationSummary} 
dt <- getTargetedGenesTable(queryRegions = queryRegions, 
                           txdbFeatures = txdbFeatures)
dt <- dt[order(transcripts, decreasing = TRUE)][1:100]

if(params$printProcessedTables == TRUE) {
  write.table(x = dt, file='getTargetedGenesTable.tsv', quote = FALSE, sep = '\t', row.names = FALSE)
} 

DT::datatable(dt, 
          extensions = 'FixedColumns', filter = 'bottom',
          options = list(fixedColumns = TRUE, scrollX = TRUE)
          )
tableCount <- tableCount + 1

```

```{r query_gene_types_figcap, results='asis', eval=params$annotationSummary}
cat("## Distribution of query regions in the genome grouped by gene types\n")
cat("**Figure",figureCount,":** The number of query regions that overlap different kinds of gene types are counted. The 'x' axis denotes the types of genes included in the analysis and the 'y' axis denotes the percentage of query regions (out of total number of query regions denoted with 'n') that overlap at least one genomic interval that host the corresponding gene type. If the query regions don't overlap any known genes, they are classified as 'Unknown'.\n")
```

```{r query_gene_types, eval=params$annotationSummary}
biotype_col <- grep('gene_biotype', colnames(overlaps.dt), value = T)
df <- overlaps.dt[,length(unique(overlappingQuery)), by = biotype_col]
colnames(df) <- c("feature", "count")
df$percent <- round(df$count / length(queryRegions) * 100, 1)
df <- df[order(count, decreasing = TRUE)]
p <- plot_ly(data = df, 
             type = "bar",
             x = feature,
             y = percent,
             text = paste("count:", count), color=feature)
layout(p = p, 
       margin = list(l=100, r=100, b=150), 
       xaxis = list(showticklabels = TRUE,  tickangle = 90), 
       yaxis = list(title = paste("percentage of query regions,", 
                                  "n =", length(queryRegions))))
figureCount <- figureCount + 1
```


```{r tssCoverage_figcap, results='asis'}
cat("## Coverage profile of query regions at/around Transcription Start/End Sites\n")
cat("**Figure",figureCount,":** The depth of coverage of query regions at and around Transcription Start/End Sites\n")
```

```{r tssCoverage}
tss <- getTSSCoverage(queryRegions = queryRegions, transcripts = txdbFeatures$transcripts, flankSize = 1000)
tes <- getTESCoverage(queryRegions = queryRegions, transcripts = txdbFeatures$transcripts, flankSize = 1000)
p <- subplot(
  plot_ly(data = tss, x = bases, y = coverage),
  plot_ly(data = tes, x = bases, y = coverage),
  margin = 0.05
) %>% layout (xaxis = list(title = 'Distance (bp) to TSS'), 
              xaxis2 = list(title = 'Distance (bp) to TES'), showlegend = FALSE) 
layout(p) 
figureCount <- figureCount + 1
```


```{r coverageprofilelist_figcap, results='asis', eval=params$annotationSummary}
cat("## Coverage profile of query regions across the length of different gene features\n")
cat("**Figure",figureCount,":** The query regions are overlaid with the genomic coordinates of features. Each entry corresponding to a feature is divided into 100 bins of equal length and for each bin the number of query regions that cover the corresponding bin is counted. Features shorter than 100bp are excluded. Thus, a coverage profile is obtained based on the distribution of the query regions. The strandedness of the features are taken into account. The coverage profile is plotted in the 5' to 3' direction.\n")
```

```{r coverageprofilelist, eval=params$annotationSummary}
covList <- calculateCoverageProfileList(queryRegions = queryRegions, 
                                       targetRegionsList = txdbFeatures, 
                                       sampleN = 10000)

df <- do.call('cbind', covList)
df <- df[,!grepl(colnames(df), pattern = '*.bins')]
df$bins <- c(1:100)
colnames(df) <- gsub(pattern = ".coverage", replacement = "", x = colnames(df))
mdf <- reshape2::melt(df, id.vars = c('bins'))
colnames(mdf) <- c('bins', 'feature', 'coverage')
p = plot_ly(data = mdf, x = bins, y = coverage, color = feature)
layout(p)

figureCount <- figureCount + 1
```

```{r plot_motif_results_figcap, results='asis', eval=params$motifAnalysis}
cat('# motifRG analysis results \n')
```

```{r motif_analysis, results='hide', eval=params$motifAnalysis}

motifResults <- runMotifRG(queryRegions = queryRegions, 
                           genomeVersion = params$genomeVersion, 
                           motifN = 4)
if (length(motifResults$motifs) > 0) {
  provideMotifSummary <- TRUE
}
```

```{r motif_analysis_plot_figcap, results='asis', eval=params$motifAnalysis}
cat('## Top motifs discovered using motifRG \n')
```

```{r motif_analysis_plot, results='asis', eval=params$motifAnalysis}
if (length(motifResults$motifs) > 0) {
  cat("**Figure-",figureCount,":** Top motifs discovered in the sequences of the query regions\n")
  par(mfrow = c(2,2), mar = c(2,2,2,2))
  for (i in 1:length(motifResults$motifs)) {
    motifPattern <- motifResults$motifs[[i]]@pattern
    motifRG::plotMotif(match = motifResults$motifs[[i]]@match$pattern, 
                       main = paste0('Motif-',i,': ',motifPattern),
                       entropy = T)
  }
  figureCount <- figureCount + 1
} else {
  cat('**Could not detect any motif enrichment in the given query regions**\n')
}
```

```{r summarize_stats_tablecap, results='asis', eval=provideMotifSummary}
cat('## motifRG motif discovery statistics \n')
cat("**Table",tableCount,":** motifRG motif discovery statistics. fg: foreground; bg: background; hits: number of motif hits; seq: number of sequences with motifs; frac: fraction of sequences that contain the motif compared to the all sequences; ratio: ratio of foreground motif fraction versus background motif fraction\n")
```

```{r motif_analysis_table, eval=provideMotifSummary}
summary <- getMotifSummaryTable(motifResults)
DT::datatable(summary, 
          extensions = 'FixedColumns', filter = 'bottom',
          options = list(fixedColumns = TRUE, scrollX = TRUE)
          )
tableCount <- tableCount + 1
```

```{r GO_section_header, results='asis', eval=params$goAnalysis}
cat("# GO Term Analysis Results\n")
```


```{r GO analysis, results='hide', eval=params$goAnalysis}

#run TopGO
goBP <- runTopGO(ontology = 'BP', 
                      species = params$species, 
                      backgroundGenes = backgroundGenes, 
                      targetedGenes = targetedGenes)

goMF <- runTopGO(ontology = 'MF', 
                      species = params$species, 
                      backgroundGenes = backgroundGenes, 
                      targetedGenes = targetedGenes)

goCC <- runTopGO(ontology = 'CC', 
                      species = params$species, 
                      backgroundGenes = backgroundGenes, 
                      targetedGenes = targetedGenes)

if (!is.null(goBP)) { goSummaryBP <- TRUE }
if (!is.null(goMF)) { goSummaryMF <- TRUE }
if (!is.null(goCC)) { goSummaryCC <- TRUE }
```

```{r goBP_tabcap, results='asis', eval=goSummaryBP}
cat("## Biological Processes\n")
cat("**Table",tableCount,":** Significant Biological Process GO terms (bonferroni < 0.1) enriched for genes that overlap query regions\n")  
goBP <- goBP[order(goBP$foldEnrichment, decreasing = TRUE),]
rownames(goBP) <- goBP$GO.ID
goBP <- subset(goBP, select = -c(bh, GO.ID))
DT::datatable(goBP[goBP$bonferroni < 0.1,], 
          extensions = 'FixedColumns', filter = 'bottom',
          options = list(fixedColumns = TRUE, scrollX = TRUE)
          )

tableCount <- tableCount + 1
```

```{r goMF_tabcap, results='asis', eval=goSummaryMF}
cat("## Molecular Functions\n")
cat("**Table",tableCount,":** Significant Molecular Function GO terms (bonferroni < 0.1) enriched for genes that overlap query regions\n")  
goMF <- goMF[order(goMF$foldEnrichment, decreasing = TRUE),]
rownames(goMF) <- goMF$GO.ID
goMF <- subset(goMF, select = -c(bh, GO.ID))

DT::datatable(goMF[goMF$bonferroni < 0.1,], 
          extensions = 'FixedColumns', filter = 'bottom',
          options = list(fixedColumns = TRUE, scrollX = TRUE)
          )
tableCount <- tableCount + 1
```

```{r goCC_tabcap, results='asis', eval=goSummaryCC}
cat("## Cellular Compartments\n")
cat("**Table",tableCount,":** Significant Cellular Compartment GO terms (bonferroni < 0.1) enriched for genes that overlap query regions\n")  
goCC <- goCC[order(goCC$foldEnrichment, decreasing = TRUE),]
rownames(goCC) <- goCC$GO.ID
goCC <- subset(goCC, select = -c(bh, GO.ID))

DT::datatable(goCC[goCC$bonferroni < 0.1,], 
          extensions = 'FixedColumns', filter = 'bottom',
          options = list(fixedColumns = TRUE, scrollX = TRUE)
          )

tableCount <- tableCount + 1
```

```{r GSEA_msigdb_tabcap, results='asis', eval=params$msigdbAnalysis}
cat("# Gene Set Enrichment Analysis Results\n")
cat("**Table",tableCount,":** Significant MSigDB Gene Sets (bonferroni < 0.1) enriched for genes that overlap query regions\n")
```

```{r msigdb_analysis, eval=params$msigdbAnalysis}
if (params$species != 'human') {
  targetMSIGDB <- createOrthologousMsigdbDataset(refMsigdbFilePath = params$msigdb, 
                                                refGenomeVersion = 'hg19', 
                                                targetGenomeVersion = params$genomeVersion)
  msigdbOutFile <- paste(params$genomeVersion, 'msigdb.gmt', sep='.')
  
  printMsigdbDataset(dataset = targetMSIGDB, 
                     outputFilename = msigdbOutFile)
  msigDB <- parseMsigdb(filePath = msigdbOutFile)
} else if (params$msigdb == 'testdata') {
  data(msigDB)
} else {
  msigDB <- parseMsigdb(filePath = params$msigdb)
}

msigdbResults <- runMSIGDB(msigDB = msigDB,
                           backgroundGenes = backgroundGenes, 
                           targetedGenes = targetedGenes)
rownames(msigdbResults) = gsub(pattern = '_', replacement = ' ', x = rownames(msigdbResults))
msigdbResults = msigdbResults[order(msigdbResults$foldEnrichment, decreasing = TRUE),]
msigdbResults <- subset(msigdbResults, select = -c(BH))

DT::datatable(msigdbResults[msigdbResults$bonferroni < 0.1,], 
          extensions = 'FixedColumns', filter = 'bottom',
          options = list(fixedColumns = TRUE, scrollX = TRUE)
          )
```


# Acknowledgements

RCAS is developed by
[Dr. Altuna Akalin](http://bioinformatics.mdc-berlin.de/team.html#altuna-akalin-phd)
(head of the Scientific Bioinformatics Platform),
[Dr. Bora Uyar](http://bioinformatics.mdc-berlin.de/team.html#bora-uyar-phd)
(Bioinformatics Scientist),
[Dr. Dilmurat Yusuf](http://bioinformatics.mdc-berlin.de/team.html#dilmurat-yusuf-phd)
(Bioinformatics Scientist) and 
[Ricardo Wurmus](http://bioinformatics.mdc-berlin.de/team.html#ricardo-wurmus)
(System Administrator) at the Berlin Institute of Medical Systems Biology
([BIMSB](https://www.mdc-berlin.de/13800178/en/bimsb))
at the Max-Delbrueck-Center for Molecular Medicine
([MDC](https://www.mdc-berlin.de)) in Berlin.

RCAS is developed as a bioinformatics service as part of
the [RNA Bioinformatics Center](http://www.denbi.de/index.php/rbc),
which is one of the eight centers of
the German Network for Bioinformatics Infrastructure
([de.NBI](http://www.denbi.de/)).  

