---
title: "The RNA Centric Analysis System Report"
---

```{r load_libraries, results='hide', message=FALSE, warning=FALSE, echo=FALSE}
library('data.table')
library('DT')
library('plotly')
library('genomation')
library('rtracklayer')
library('GenomicFeatures')
library('knitr')
library('motifRG')
library('reshape2')
```


```{r get_inputs, echo=FALSE, eval=TRUE}
args <- commandArgs(TRUE)
work_dir = args[1] #path to working directory - necessary to set root.dir
peaks_file = args[2] #'/home/buyar/projects/RCAS/test/PARCLIP_AGO1234_Hafner2010a_hg19_xaa.bed'
gff_rds = args[3] #'/data/akalin/Base/Annotation/GenomeAnnotation/hg19/gencode/gencode.v19.annotation.gff3.granges.rds' #RDS session file as output from import gff function of rtracklayer library
txdb_file = args[4] #txdb file generated based on input gff file
go_bp = args[5] ##go enrichment results for biological processes
go_mf = args[6] ##go enrichment results for molecular functions
go_cc = args[7] ##go enrichment results for sub-cellular locations
msigdb = args[8] ##msigdb enrichment results 
coverage_profile_option = args[9] ##option to run or not run the coverage profile calculations
motif_results = args[10] # motifRG results in rds format e.g. <sample>.motif.results.rds
peak_seqs_fasta = args[11] # fasta sequences of peaks, output of rcas.motif.R <sample>.foreground.fasta 
figure_count = 1
table_count = 1
```


```{r define_modules_to_run, echo=FALSE, warning=FALSE, eval=TRUE}
##based on the input options; define boolean parameters that are used in corresponding chunks to determine if chunk should be evaluated or not. 
run_anot = TRUE #basic annotation summary is always run
eval_query_gene_types = FALSE #depends on run_anot; 
run_go_bp = ifelse(go_bp == 'NOT_RUN', FALSE, TRUE)
run_go_mf = ifelse(go_mf == 'NOT_RUN', FALSE, TRUE)
run_go_cc = ifelse(go_cc == 'NOT_RUN', FALSE, TRUE)
run_go = run_go_bp || run_go_mf || run_go_cc 
run_msigdb = ifelse(msigdb == 'NOT_RUN', FALSE, TRUE)
run_coverage_profile = ifelse(coverage_profile_option == 'NOT_RUN', FALSE, TRUE)
run_motif = ifelse(motif_results == 'NOT_RUN', FALSE, TRUE)
provide_motif_summary = FALSE #depends on run_anot and run_motif being TRUE 
```


```{r setup, echo=FALSE}
opts_knit$set(root.dir=work_dir)
```

```{r read_gff_rds, echo=FALSE, warning=FALSE, message=FALSE, eval=run_anot} 
#prepare datasets that will be used in later chunks 
gff = readRDS(gff_rds) #GRanges object 
```

```{r get_txdb_features ,echo=FALSE, warning=FALSE, message=FALSE, eval=run_anot}
txdb = loadDb(txdb_file)

txdb_transcripts = transcripts(txdb) #has tx_name
txdb_transcripts$gene_name = gff[match(txdb_transcripts$tx_name, gff$transcript_id)]$gene_name

txdb_exons = unlist(exonsBy(txdb, by="tx", use.names=TRUE)) #row names are transcript_ids
txdb_exons$tx_name = names(txdb_exons)
txdb_exons$gene_name = gff[match(names(txdb_exons), gff$transcript_id)]$gene_name

txdb_introns = unlist(intronsByTranscript(txdb, use.names=TRUE))#row names are transcript_ids
txdb_introns$tx_name = names(txdb_introns)
txdb_introns$gene_name = gff[match(names(txdb_introns), gff$transcript_id)]$gene_name

txdb_promoters = promoters(txdb) #has tx_name
txdb_promoters$gene_name = gff[match(txdb_promoters$tx_name, gff$transcript_id)]$gene_name

txdb_five_UTRs = unlist(fiveUTRsByTranscript(txdb,use.names=TRUE))#row names are transcript_ids
txdb_five_UTRs$tx_name = names(txdb_five_UTRs)
txdb_five_UTRs$gene_name = gff[match(names(txdb_five_UTRs), gff$transcript_id)]$gene_name

txdb_three_UTRs = unlist(threeUTRsByTranscript(txdb,use.names=TRUE))#row names are transcript_ids
txdb_three_UTRs$tx_name = names(txdb_three_UTRs)
txdb_three_UTRs$gene_name = gff[match(names(txdb_three_UTRs), gff$transcript_id)]$gene_name

txdb_cds = unlist(cdsBy(txdb, by="tx",use.names=TRUE))#row names are transcript_ids
txdb_cds$tx_name = names(txdb_cds)
txdb_cds$gene_name = gff[match(names(txdb_cds), gff$transcript_id)]$gene_name

#define a function that can be used to summarize a given set of peaks 
summarize_peaks = function(peaks){
  tr = length(unique(queryHits(findOverlaps(peaks, txdb_transcripts))))
  pr = length(unique(queryHits(findOverlaps(peaks, txdb_promoters))))
  futr = length(unique(queryHits(findOverlaps(peaks, txdb_five_UTRs))))
  ex = length(unique(queryHits(findOverlaps(peaks, txdb_exons))))
  cds = length(unique(queryHits(findOverlaps(peaks, txdb_cds))))
  inr = length(unique(queryHits(findOverlaps(peaks, txdb_introns))))
  tutr = length(unique(queryHits(findOverlaps(peaks, txdb_three_UTRs))))
  res = c(tr, pr, futr, ex, cds, inr, tutr)
  df = data.frame(freq = res, features = c('transcripts', 'promoters', "5'UTR", 'exons', 'cds', 'intron', "3'UTR"))
  return(df)
}

```

```{r read_peaks,echo=FALSE, warning=FALSE, message=FALSE, eval=run_anot}
peaks = import(peaks_file, format = "bed")
seqlevelsStyle(peaks) = 'UCSC'
```

```{r plot_motif_results_figcap, echo=FALSE, results='asis', eval=run_motif}
cat('# motifRG analysis results \n')
cat('## Top motifs discovered using motifRG \n')
```

```{r plot_motif_results, results='asis', echo=FALSE, eval=run_motif}
motif = readRDS(motif_results)
peak_seqs = readDNAStringSet(peak_seqs_fasta)
if(length(motif$motifs) > 0){
  if(run_anot==TRUE) {provide_motif_summary = TRUE}
  cat(paste0("**Figure-",figure_count,":** Top motifs discovered in the sequences of the query regions"))
  par(mfrow=c(3,2), mar=c(2,2,2,2))
  for (i in 1:length(motif$motifs)){
    motif_pattern = motif$motifs[[i]]@pattern
    plotMotif(motif$motifs[[i]]@match$pattern, main=paste0('Motif-',i,': ',motif_pattern), entropy = T)
  }
  figure_count = figure_count + 1
}else{
  cat('**Could not detect any motif enrichment in the given query regions**\n')
}
```

```{r summarize_stats_tablecap, echo=FALSE, results='asis', eval=provide_motif_summary}
cat('## motifRG motif discovery statistics \n')
cat("**Table",table_count,":** motifRG motif discovery statistics. fg: foreground; bg: background; hits: number of motif hits; seq: number of sequences with motifs; frac: fraction of sequences that contain the motif compared to the all sequences; ratio: ratio of foreground motif fraction versus background motif fraction\n")
```

```{r summarize_stats_table, echo=FALSE, warning=FALSE, eval=provide_motif_summary}
motif = readRDS(motif_results)
df = summaryMotif(motif$motifs, motif$category)
df$scores = round(df$scores, 1)
df$ratio = round(df$ratio, 1)
datatable(df[,-c(2)], extensions = 'FixedColumns',
  options = list(
    dom = 't',
    scrollX = TRUE,
    scrollCollapse = TRUE
  ))
table_count = table_count + 1
```

```{r summarize_motif_hits_figcap, echo=FALSE, results='asis', eval=provide_motif_summary}
cat('## Genomic features overlapping with discovered motifs \n')
cat("**Figure",figure_count,":** The number of motif containing query regions in different gene features.\n")

```

```{r summarize_motif_hits, echo=FALSE, warning=FALSE, eval=provide_motif_summary}

motif = readRDS(motif_results)
peak_seqs = readDNAStringSet(peak_seqs_fasta) 
peak_ids = which(motif$category==1) #which ids correspond to the foreground seqs (peaks)

#the sequence of features must be the same as defined in summarize_peaks function
motif_summary = data.frame(features = c('transcripts', 'promoters', "5'UTR", 'exons', 'cds', 'intron', "3'UTR"))

for (i in 1:length(motif$motifs)){
  match_seqid = motif$motifs[[i]]@match$seq.id #found sequence ids of all query regions (both foreground (peaks) and background)
  match_peaks = peaks[match_seqid[match_seqid %in% peak_ids]] #found peaks that match the current motif
  motif_pattern = motif$motifs[[i]]@pattern
  #now do summary stats by overlap operations 
  res = summarize_peaks(match_peaks)
  motif_summary = cbind(motif_summary, res$freq)
  colnames(motif_summary)[length(colnames(motif_summary))] = motif_pattern 
}

m = melt.data.table(data.table(motif_summary), id.vars = c('features'))
p = plot_ly(data = m, x = variable, y = value, color = features, type='bar')
layout(p, xaxis = list(title="Discovered Motif Patterns"), yaxis=list(title=paste0('Number of Peaks containing the motif')))
figure_count = figure_count + 1

```


```{r overlap_gene_features_figcap, echo=FALSE, results='asis', eval=run_anot}
cat('# Distribution of query regions in genomic annotation features \n')
cat('## Distribution of query regions across gene features\n')

cat("**Figure",figure_count,":** The number of query regions that overlap different kinds of gene features are counted. The 'y' axis denotes the types of gene features included in the analysis and the 'x' axis denotes the percentage of query regions (out of total number of query regions denoted with 'n') that overlap at least one genomic interval that host the corresponding feature. Notice that the sum of the percentage values for different features don't add up to 100%, because some query regions may overlap multiple kinds of features \n")

```

```{r overlap_gene_features, echo=FALSE, warning=FALSE, message=FALSE, eval=run_anot}
res = summarize_peaks(peaks)

p <- plot_ly(res, x=features, y=freq, type = 'bar')
layout(p, xaxis = list(title="Gene Features"), yaxis=list(title=paste0('Number of Peaks (out of ',length(peaks),')')))

figure_count = figure_count + 1
```

```{r find_overlaps_gff_peaks, echo=FALSE, warning=FALSE, message=FALSE, eval=run_anot}
o = findOverlaps(gff, peaks) #find all overlapping pairs of intervals between the gff features and BED file (peaks)
o.gff = gff[queryHits(o)]
o.peak = as.data.frame(peaks[subjectHits(o)])
o.gff$overlapping_peak = paste(o.peak$seqnames, o.peak$start, o.peak$end, o.peak$strand, sep=':')
overlaps.gr = o.gff
overlaps.dt = data.table(as.data.frame(o.gff)) ##convert granges object into data table

if (any(grepl('biotype', colnames(overlaps.dt)))){
  eval_query_gene_types = TRUE
}
```

```{r query_gene_types_figcap, echo=FALSE, results='asis', eval=eval_query_gene_types}
cat("## Distribution of query regions in the genome grouped by gene types\n")
cat("**Figure",figure_count,":** The number of query regions that overlap different kinds of gene types are counted. The 'x' axis denotes the types of genes included in the analysis and the 'y' axis denotes the percentage of query regions (out of total number of query regions denoted with 'n') that overlap at least one genomic interval that host the corresponding gene type. If the query regions don't overlap any known genes, they are classified as 'Unknown'.\n")
```

```{r query_gene_types, echo=FALSE, warning=FALSE, message=FALSE, eval=eval_query_gene_types}
biotype_col = grep('biotype', colnames(overlaps.dt), value=T)
df = overlaps.dt[,length(unique(overlapping_peak)), by=biotype_col]
colnames(df) = c("feature", "count")
df$percent = round(df$count / length(unique(overlaps.dt$overlapping_peak)) * 100, 1)
df = df[order(count, decreasing = TRUE)]
p = plot_ly(df, type="bar", x = feature, y = percent, text=paste("count:", count), color=feature)
layout(p, margin = list(l=100, r=100, b=150), xaxis = list(showticklabels = TRUE,  tickangle = 90), yaxis = list(title = paste("percentage of query regions,", "n =",length(unique(overlaps.dt$overlapping_peak)))))
figure_count = figure_count + 1
```

```{r chromosomes_gene_features_figcap, echo=FALSE, results='asis', eval=run_anot}
cat("## Distribution of query regions across the chromosomes grouped by gene features\n")
cat("**Figure",figure_count,":** The number of query regions that overlap different chromosomes are counted. For each chromosome, the frequency of query regions are further split into groups based on the gene features the query overlaps with. The 'x' axis denotes the chromosomes included in the analysis and the 'y' axis denotes the frequency of overlaps.\n")
```

```{r chromosomes_gene_features,  echo=FALSE, warning=FALSE, message=FALSE, eval=run_anot}
df = overlaps.dt[,length(unique(overlapping_peak)), by=c('seqnames', 'type')]
colnames(df) = c('seqnames', 'type', 'count')
df = df[order(seqnames)]
p = plot_ly(df, type="bar", x = seqnames, y = count, text=paste("count:", count), color=type)
layout(p, margin = list(l=100, r=100, b=150), xaxis = list(showticklabels = TRUE,  tickangle = 90), yaxis = list(title = paste("Number of query regions,", "n =",length(unique(overlaps.dt$overlapping_peak)))))

figure_count = figure_count + 1
```


```{r get_table_of_genes_tabcap, echo=FALSE, results='asis', eval=run_anot}
cat("## Interactive table of genes that overlap query regions\n")
cat("**Table",table_count,":** Interactive table of genes (top 500) that overlap query regions, grouped by gene features such as introns, exons, UTRs, etc.\n")
```

```{r get_table_of_genes, echo=FALSE, message=FALSE, warning=FALSE, eval=run_anot}

process_hits = function(peaks, tx, type){
  o = findOverlaps(peaks, tx)
  o.peak = as.data.frame(peaks[queryHits(o)])
  o.tx = tx[subjectHits(o)]  
  o.tx$overlapping_peak = paste(o.peak$seqnames, o.peak$start, o.peak$end, o.peak$strand, sep=':')
  dt = data.table(data.frame(tx_name = o.tx$tx_name, overlapping_peak = o.tx$overlapping_peak))
  summary = dt[,length(unique(overlapping_peak)), by='tx_name']
  colnames(summary) = c('tx_name', type)
  return(summary)
}

tr = process_hits(peaks, txdb_transcripts, 'transcript_total')
pr = process_hits(peaks, txdb_promoters, 'promoter')
futr = process_hits(peaks, txdb_five_UTRs, "5'UTR")
ex = process_hits(peaks, txdb_exons, 'exon')
cds = process_hits(peaks, txdb_cds, 'cds')
inr = process_hits(peaks, txdb_introns, 'intron')
tutr = process_hits(peaks, txdb_three_UTRs, "3'UTR")

tbls <- list(pr, futr, ex, cds, inr, tutr, tr)
tbls = lapply(tbls, function(i) setkey(i, tx_name))
merged <- Reduce(function(...) merge(..., all = T), tbls)
merged$gene_name = gff[match(merged$tx_name, gff$transcript_id)]$gene_name
merged = merged[order(transcript_total, decreasing=T)]
datatable(merged[1:500], filter = 'top', extensions = 'FixedColumns',
  options = list(
    dom = 't',
    scrollX = TRUE,
    scrollCollapse = TRUE
  ))
table_count = table_count + 1
```


```{r coverage_profiles_section_header, echo=FALSE, results='asis', eval=run_coverage_profile}
cat("# Coverage Profiles\n")
```

```{r coverage_profiles_read_input, echo=FALSE, warning=FALSE, message=FALSE, eval=run_coverage_profile}

calculate_coverage = function (peaks, windows, type){
  #windows = windows[width(windows) > 100] #remove windows shorter than 100 bp
  if (length(windows) > 10000){
    windows = windows[sample(length(windows), 10000)] #if there are more than 10000 windows; downsample to 10000 
  }
  sm = ScoreMatrixBin(target = peaks, windows = windows, bin.num = 100, strand.aware = TRUE)
  mdata = as.data.frame(colSums(sm))
  mdata$bins = c(1:100)
  colnames(mdata) = c('coverage', 'bins')
  p = plot_ly(mdata, x = bins, y = coverage)
  p %>%  
  add_trace(y = fitted(loess(coverage ~ as.numeric(bins)))) %>%
  layout(title = paste("Coverage along", type, "(5' -> 3' direction)", sep=" "), showlegend = FALSE, margin = list(l= 50, r=50, b=50, t=50))
}
```

```{r coverage_profiles_transcripts_figcap, echo=FALSE, results='asis', eval=run_coverage_profile}
cat("## Coverage profile of query regions across the length of transcripts\n")
cat("**Figure",figure_count,":** The query regions are overlaid with the genomic coordinates of transcripts. The transcripts are divided into 100 bins of equal length and for each bin the number of query regions that cover the corresponding bin is counted. Transcripts shorter than 100bp are excluded. Thus, a coverage profile of the transcripts is obtained based on the distribution of the query regions. The strandedness of the transcripts are taken into account. The coverage profile is plotted in the 5' to 3' direction.\n")
```

```{r coverage_profiles_transcripts, echo=FALSE, warning=FALSE, message=FALSE, eval=run_coverage_profile}
calculate_coverage (peaks, windows = txdb_transcripts, "transcripts") 
figure_count = figure_count + 1
```

```{r coverage_profiles_exons_figcap, echo=FALSE, results='asis', eval=run_coverage_profile}
cat("## Coverage profile of query regions across the length of Exons\n")
cat("**Figure",figure_count,":** The query regions are overlaid with the genomic coordinates of each exon of each transcript. The exons are divided into 100 bins of equal length and for each bin the number of query regions that cover the corresponding bin is counted. Exons shorter than 100bp are excluded. Thus, a coverage profile of the exons is obtained based on the distribution of the query regions. The strandedness of the exons are taken into account. The coverage profile is plotted in the 5' to 3' direction.\n")
```

```{r coverage_profiles_exons, echo=FALSE, warning=FALSE, message=FALSE, eval=run_coverage_profile}
calculate_coverage (peaks, windows = txdb_exons, "exons")
figure_count = figure_count + 1
```

```{r coverage_profiles_exon-introns_figcap, echo=FALSE, results='asis', eval=run_coverage_profile}
cat("## Coverage profile of query regions across the 100 bp region centered on exon-intron junctions\n")
cat("**Figure",figure_count,":** The query regions are overlaid with the genomic coordinates of each exon-intron junction of each transcript. The junction comprises of a 50 bp region of an exon and 50 bp region of its neighboring intron. Exon-intron junctions are divided into 100 bins of equal length and for each bin the number of query regions that cover the corresponding bin is counted. Exons shorter than 100bp are excluded. Thus, a coverage profile of the exon-intron junctions is obtained based on the distribution of the query regions. The strandedness of the exons are taken into account. The coverage profile is plotted in the 5' to 3' direction.\n")
```

```{r coverage_profiles_exon-introns, echo=FALSE, warning=FALSE, message=FALSE, eval=run_coverage_profile}
myexons = txdb_exons[width(txdb_exons) > 100] #remove exons that are shorter than 100 bp
up_exon = flank(myexons, 50, start=TRUE, both=TRUE) #50 bp from upstream intron 50 bp from exon 
down_exon = flank(myexons, 50, start=FALSE, both=TRUE) #50 bp from downstream intron, 50 bp from exon
exon_intron_junctions = c(up_exon, down_exon)
calculate_coverage (peaks, windows = exon_intron_junctions, "exon-intron junctions")
figure_count = figure_count + 1
```

```{r coverage_profiles_introns_figcap, echo=FALSE, results='asis', eval=run_coverage_profile}
cat("## Coverage profile of query regions across the length of introns\n")
cat("**Figure",figure_count,":** The query regions are overlaid with the genomic coordinates of each intron of each transcript. The introns are divided into 100 bins of equal length and for each bin the number of query regions that cover the corresponding bin is counted. Introns shorter than 100bp are excluded. Thus, a coverage profile of the introns is obtained based on the distribution of the query regions. The strandedness of the introns are taken into account. The coverage profile is plotted in the 5' to 3' direction.\n")
```

```{r coverage_profiles_introns, echo=FALSE, warning=FALSE, message=FALSE, eval=run_coverage_profile}
calculate_coverage (peaks, windows = txdb_introns, "introns")
figure_count = figure_count + 1
```


```{r coverage_profiles_promoters_figcap, echo=FALSE, results='asis', eval=run_coverage_profile}
cat("## Coverage profile of query regions across the promoter regions\n")
cat("**Figure",figure_count,":** The query regions are overlaid with the genomic coordinates of each promoter region of each transcript. The promoter region is defined as the region spanning from 2000bp upstream of the transcription start site and the first 200bp region after the transcription start site. The promoters are divided into 100 bins of equal length and for each bin the number of query regions that cover the corresponding bin is counted. Thus, a coverage profile of the promoters is obtained based on the distribution of the query regions. The strandedness of the promoters are taken into account. The coverage profile is plotted in the 5' to 3' direction.\n")
```

```{r coverage_profiles_promoters, echo=FALSE, warning=FALSE, message=FALSE, eval=run_coverage_profile}
calculate_coverage (peaks, windows = txdb_promoters, "promoters")
figure_count = figure_count + 1
```


```{r coverage_profiles_fiveUTRs_figcap, echo=FALSE, results='asis', eval=run_coverage_profile}
cat("## Coverage profile of query regions across the length of 5' UTRs\n")
cat("**Figure",figure_count,":** The query regions are overlaid with the genomic coordinates of each 5' UTR region of each transcript. The 5' UTR regions are divided into 100 bins of equal length and for each bin the number of query regions that cover the corresponding bin is counted. Thus, a coverage profile of the 5' UTR regions is obtained based on the distribution of the query regions. The strandedness of the promoters are taken into account. The coverage profile is plotted in the 5' to 3' direction.\n")
```

```{r coverage_profiles_fiveUTRs, echo=FALSE, warning=FALSE, message=FALSE, eval=run_coverage_profile}
calculate_coverage (peaks, windows = txdb_five_UTRs, "5' UTRs")
figure_count = figure_count + 1
```


```{r coverage_profiles_threeUTRs_figcap, echo=FALSE, results='asis', eval=run_coverage_profile}
cat("## Coverage profile of query regions across the length of 3' UTRs\n")
cat("**Figure",figure_count,":** The query regions are overlaid with the genomic coordinates of each 3' UTR region of each transcript. The 3' UTR regions are divided into 100 bins of equal length and for each bin the number of query regions that cover the corresponding bin is counted. Thus, a coverage profile of the 3' UTR regions is obtained based on the distribution of the query regions. The strandedness of the promoters are taken into account. The coverage profile is plotted in the 5' to 3' direction.\n")
```

```{r coverage_profiles_threeUTRs, echo=FALSE, warning=FALSE, message=FALSE, eval=run_coverage_profile}
calculate_coverage (peaks, windows = txdb_three_UTRs, "3' UTRs")
figure_count = figure_count + 1
```

```{r GO_section_header, results='asis', echo=FALSE, eval=run_go}
cat("# GO term and Pathway Enrichment Results\n")
```

```{r GO_BP_tabcap, results='asis', echo=FALSE, eval=run_go_bp}
cat("## GO Term Enrichment Results for Biological Processes\n")
cat("**Table",table_count,":** Significant Biological Process GO terms (BH < 0.1) enriched for genes that overlap query regions\n")  
```

```{r GO_BP, results='asis', warning=FALSE, echo=FALSE, eval=run_go_bp}
go = fread(go_bp)
results = go[bh < 0.1]
if (nrow(results) > 0){
  datatable(results, filter = 'top', extensions = 'FixedColumns',
  options = list(
    dom = 't',
    scrollX = TRUE,
    scrollCollapse = TRUE
  ))
}else{
    cat("**WARNING: Couldn't detect any enrichment/depletion of GO terms (BH < 0.001) related to biological processes**\n")
}
table_count = table_count + 1
```


```{r GO_MF_tabcap, results='asis', echo=FALSE, eval=run_go_mf}
cat("## GO Term Enrichment Results for Molecular Functions\n")
cat("**Table",table_count,":** Significant Molecular Function GO terms (BH < 0.1) enriched for genes that overlap query regions\n")
```

```{r GO_MF, results='asis', warning=FALSE,  echo=FALSE, eval=run_go_mf}
go = fread(go_mf)
results = go[bh < 0.1]
if (nrow(results) > 0){
  datatable(results,   filter = 'top', extensions = 'FixedColumns',
  options = list(
    dom = 't',
    scrollX = TRUE,
    scrollCollapse = TRUE
  ))
}else{
  cat("**WARNING: Couldn't detect any enrichment/depletion of GO terms (BH < 0.1) related to molecular functions**\n")
}
table_count = table_count + 1
```


```{r GO_CC_tabcap, results='asis', echo=FALSE, eval=run_go_cc}
cat("## GO Term Enrichment Results for Cellular Compartments\n")
cat("**Table",table_count,":** Significant Cellular Compartment GO terms (BH < 0.1) enriched for genes that overlap query regions\n")
```

```{r GO_CC, results='asis', warning=FALSE,  echo=FALSE, eval=run_go_cc}
go = fread(go_cc)
results = go[bh < 0.1]
if (nrow(results) > 0){
  datatable(results,  filter = 'top', extensions = 'FixedColumns',
  options = list(
    dom = 't',
    scrollX = TRUE,
    scrollCollapse = TRUE
  ))
}else{
  cat("**WARNING: Couldn't detect any enrichment/depletion of GO terms related to subcellular localisation**\n")
}
table_count = table_count + 1
```

```{r GSEA_msigdb_tabcap, results='asis', echo=FALSE, eval=run_msigdb}
cat("## Gene Set Enrichment Results based on MSigDB\n")
cat("**Table",table_count,":** Significant MSigDB Gene Sets (BH < 0.1) enriched for genes that overlap query regions\n")
```

```{r GSEA_msigdb, results='asis', warning=FALSE,  echo=FALSE, eval=run_msigdb}
msigdb = fread(msigdb)
msigdb = msigdb[order(pval)]
if (nrow(msigdb[BH < 0.1]) > 0){
  datatable(msigdb[BH < 0.1], filter='top', extensions = 'FixedColumns',
  options = list(
    dom = 't',
    scrollX = TRUE,
    scrollCollapse = TRUE
  ))
}else{
   cat("**WARNING: Couldn't detect any enrichment/depletion of MSIGDB gene sets**\n")
}
table_count = table_count + 1
```


# Acknowledgements

RCAS is developed by [Dr. Altuna Akalin](http://bioinformatics.mdc-berlin.de/team.html#altuna-akalin-phd) (head of the Scientific Bioinformatics Platform), [Dr. Dilmurat Yusuf](http://bioinformatics.mdc-berlin.de/team.html#dilmurat-yusuf-phd) (Bioinformatics Scientist), [Dr. Bora Uyar](http://bioinformatics.mdc-berlin.de/team.html#bora-uyar-phd)  (Bioinformatics Scientist), and [Ricardo Wurmus](http://bioinformatics.mdc-berlin.de/team.html#ricardo-wurmus) (System Administrator) at the Berlin Institute of Medical Systems Biology ([BIMSB](https://www.mdc-berlin.de/13800178/en/bimsb)) at the Max-Delbrueck-Center for Molecular Medicine ([MDC](https://www.mdc-berlin.de)) in Berlin. 

RCAS is developed as a bioinformatics service as part of the [RNA Bioinformatics Center](http://www.denbi.de/index.php/rbc), which is one of the eight centers of the German Network for Bioinformatics Infrastructure ([de.NBI](http://www.denbi.de/)).  








