
---
title: "The RNA Centric Annotation System Analysis Report"
author: "Dilmurat Yusuf, Bora Uyar, Altuna Akalin"
date: "30 Nov 2015"
output: 
  html_document:
    theme: flatly
    highlight: pygments
    css: custom.css
    toc: true
    number_sections: true
    includes: 
      before_body: "header.html"
---

**RCAS** is an RNA-centric annotation system for peaks discovered by CLIP-based high-throughput sequencing methods. The annotation pipeline takes as input [a 'bed' format file](https://genome.ucsc.edu/FAQ/FAQformat.html#format1) that contains the genomic coordinates of peak regions (which is the output of CLIP-based data analysis pipelines). These peaks are overlaid with genomic features that are publicly available (or provided by the user) for the corresponding species. Then, an html/pdf based report is generated including figures and graphics that summarize the annotations associated to the peak regions. 

```{r message=FALSE, warning=FALSE, echo=FALSE}
library('data.table')
library('DT')
library('plotly')
library('dplyr')
library('genomation')
library('rtracklayer')
```


```{r, fig.width=24, fig.height=24, echo=FALSE}
args <- commandArgs(TRUE)
anot = args[1] #/home/buyar/projects/RCAS/test/PARCLIP_AGO1234_Hafner2010a_hg19_xaa.anot.tsv
peaks = args[2] #'/home/buyar/projects/RCAS/test/PARCLIP_AGO1234_Hafner2010a_hg19_xaa.bed'
gff.rds = args[3] #'/data/akalin/Base/Annotation/GenomeAnnotation/hg19/gencode/gencode.v19.annotation.gff3.rds'
go_bp = args[4] ##go enrichment results for biological processes
go_mf = args[5] ##go enrichment results for molecular functions
go_cc = args[6] ##go enrichment results for sub-cellular locations
msigdb = args[7] ##msigdb enrichment results 

mdata = fread(anot)
```

# Summary Figures

------

## Distribution of intervals across gene features
```{r echo=FALSE}
dt = mdata
df = dt[,length(gene_type), by=feature]
colnames(df) = c("feature", "count")
df = df[order(count, decreasing = TRUE)]
p = plot_ly(df, x=feature, y=count, mode = "markers", marker = list(color = count, size=50)) 
p
```

## Distribution of intervals across gene features
```{r echo=FALSE}
dt = mdata
df = dt[,length(gene_type), by=feature]
colnames(df) = c("feature", "count")
df = df[order(count, decreasing = TRUE)]
p = plot_ly(df, type="pie",labels=feature,values=count, hole=0.5, text=feature, textfont = list(size=18, color="black"))
p
```

## Distribution of intervals across RNA genes
```{r echo=FALSE}
m = grepl("RNA", mdata$gene_type)
dt = mdata[m]
df = dt[,length(feature), by=gene_type]
colnames(df) = c("feature", "count")
df = df[order(count, decreasing = TRUE)]
p = plot_ly(df, type="pie",labels=feature,values=count, hole=0.5, text=feature, textfont = list(size=18, color="black"))
p
```

## Distribution of intervals in the genome grouped by gene types 
```{r echo=FALSE}
dt = mdata[,length(feature), by=gene_type]
dt = dt[order(-V1)]
colnames(dt) = c('gene_type', 'count')

p = plot_ly(dt, x = gene_type, y = log2(count), size = count, color = count, mode = "markers")
layout(p, xaxis = list(showticklabels = TRUE, tickangle = 90))

```

## Distribution of intervals across the chromosomes grouped by gene features 
```{r echo=FALSE}
p = mdata %>% count(chromosome_id, feature) %>%
  plot_ly(x = chromosome_id, y = n, type = "bar", color = feature) %>%
  layout(xaxis = list(tickangle = 90))
p 
```

------

## Interactive table of genes that overlap peaks 
```{r, echo=FALSE}
library(DT)
features=unique(mdata$feature)
mytable = mdata[feature==features[1],length(feature),by=c('gene_name')]
colnames(mytable) = c('gene_name', features[1])

for (i in 2:length(features)){
  f = features[i]
  dt = mdata[feature==f,length(feature),by=gene_name]
  colnames(dt) = c('gene_name', f)
  mytable = merge(mytable, dt, by='gene_name', all=TRUE)
}
#mt = mytable[order(UTR, decreasing=TRUE)][1:500]

datatable(mytable, caption = 'Table 1: Interactive table of genes that overlap peaks, grouped by gene features such as introns, exons, UTRs, etc.', filter = 'top' )
```

------

# Peak Coverage Profiles

## Coverage of Peak regions across the length of transcripts
```{r echo=FALSE, warning=FALSE, message=FALSE}
gff = readRDS(gff.rds)
peaks = import(peaks, format = "bed")

tr = ScoreMatrixBin(target = peaks, windows = gff[seqnames(gff)%in% c("chr21", "chr22") & gff$type=="transcript"], bin.num = 100, strand.aware = TRUE)

ex = ScoreMatrixBin(target = peaks, windows = gff[seqnames(gff)%in% c("chr21", "chr22") & gff$type=="exon"], bin.num = 100, strand.aware = TRUE)

utr = ScoreMatrixBin(target = peaks, windows = gff[seqnames(gff)%in% c("chr21", "chr22") & gff$type=="UTR"], bin.num = 100, strand.aware = TRUE)

#pdf("peak_coverage.pdf")
par(mfrow=c(3,1))
plotMeta(tr, xlab = "transcript length (normalized to 100)")
plotMeta(ex, xlab = "exon length (normalized to 100)")
plotMeta(utr, xlab = "UTR length (normalized to 100)")
#dev.off()
```

------


# GO term and Pathway Enrichment Results

## GO Term Enrichment Results for Biological Processes
```{r echo=FALSE}
go = fread(go_bp)
datatable(go[bonferroni < 0.001], caption = 'Table 2: Significant Biological Process GO terms (FDR < 0.001) enriched for genes that overlap peaks',  filter = 'top')
```

## GO Term Enrichment Results for Molecular Functions
```{r echo=FALSE}
go = fread(go_mf)

datatable(go[bonferroni < 0.001], caption = 'Table 3: Significant Molecular Function GO terms (FDR < 0.001) enriched for genes that overlap peaks',  filter = 'top')
```

## GO Term Enrichment Results for Cellular Compartments
```{r echo=FALSE}
go = fread(go_cc)
datatable(go[bonferroni < 0.001], caption = 'Table 4: Significant Cellular Compartment GO terms (FDR < 0.001) enriched for genes that overlap peaks',  filter = 'top')
```

## Gene Set Enrichment Results based on MSigDB
```{r echo=FALSE}
msigdb = fread(msigdb)
datatable(msigdb, caption='Table 5: Significant MSigDB Gene Sets (FDR < 0.001) enriched for genes that overlap peaks', filter='top')
```





